File: backend/app/schemas/review.py
from datetime import datetime

from pydantic import BaseModel, Field


class ReviewOut(BaseModel):
    id: int
    author_name: str
    rating: int
    title: str | None
    comment: str | None
    is_verified_purchase: bool
    status: str
    helpful_count: int
    created_at: datetime
    is_own: bool = False


class RatingBreakdownBucket(BaseModel):
    rating: int
    count: int


class ReviewSummaryOut(BaseModel):
    rating_avg: float | None
    rating_count: int
    review_count: int
    breakdown: list[RatingBreakdownBucket]


class ReviewListMeta(BaseModel):
    page: int
    per_page: int
    total: int
    total_pages: int
    has_next: bool


class ReviewListOut(BaseModel):
    data: list[ReviewOut]
    meta: ReviewListMeta
    summary: ReviewSummaryOut


class ReviewCreateIn(BaseModel):
    order_item_id: int
    rating: int = Field(ge=1, le=5)
    title: str | None = Field(default=None, max_length=150)
    comment: str | None = Field(default=None, max_length=2000)


class ReviewUpdateIn(BaseModel):
    rating: int = Field(ge=1, le=5)
    title: str | None = Field(default=None, max_length=150)
    comment: str | None = Field(default=None, max_length=2000)


class ReviewEligibleItemOut(BaseModel):
    order_item_id: int
    order_number: str
    title_snapshot: str
    image_snapshot: str | None
    delivered_at: datetime | None


class ReviewEligibilityOut(BaseModel):
    can_review: bool
    eligible_items: list[ReviewEligibleItemOut]
    existing_review: ReviewOut | None = None


class AdminReviewOut(BaseModel):
    id: int
    product_id: int
    product_slug: str
    product_title: str
    author_name: str
    rating: int
    title: str | None
    comment: str | None
    is_verified_purchase: bool
    status: str
    helpful_count: int
    reported_count: int
    moderated_at: datetime | None
    created_at: datetime


class AdminReviewListOut(BaseModel):
    data: list[AdminReviewOut]
    meta: ReviewListMeta


class AdminReviewModerateIn(BaseModel):
    status: str = Field(pattern="^(approved|rejected|hidden)$")

File: frontend/src/types/review.ts
export interface Review {
  id: number
  author_name: string
  rating: number
  title: string | null
  comment: string | null
  is_verified_purchase: boolean
  status: 'pending' | 'approved' | 'rejected' | 'hidden'
  helpful_count: number
  created_at: string
  is_own: boolean
}

export interface RatingBreakdownBucket {
  rating: number
  count: number
}

export interface ReviewSummary {
  rating_avg: number | null
  rating_count: number
  review_count: number
  breakdown: RatingBreakdownBucket[]
}

export interface ReviewListMeta {
  page: number
  per_page: number
  total: number
  total_pages: number
  has_next: boolean
}

export interface ReviewListResponse {
  data: Review[]
  meta: ReviewListMeta
  summary: ReviewSummary
}

export interface ReviewEligibleItem {
  order_item_id: number
  order_number: string
  title_snapshot: string
  image_snapshot: string | null
  delivered_at: string | null
}

export interface ReviewEligibility {
  can_review: boolean
  eligible_items: ReviewEligibleItem[]
  existing_review: Review | null
}

export interface AdminReview {
  id: number
  product_id: number
  product_slug: string
  product_title: string
  author_name: string
  rating: number
  title: string | null
  comment: string | null
  is_verified_purchase: boolean
  status: 'pending' | 'approved' | 'rejected' | 'hidden'
  helpful_count: number
  reported_count: number
  moderated_at: string | null
  created_at: string
}

export interface AdminReviewListResponse {
  data: AdminReview[]
  meta: ReviewListMeta
}
